provider/pkg/await/condition/observer.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package condition
+
+import (
+	"context"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+var (
+	_ Observer = (*ObjectObserver)(nil)
+	_ observer = observer{}
+)
+
+// Observer acts on a watch.Event Source. Range is responsible for filtering
+// events to only those relevant to the Observer, and Observe optionally
+// updates the Observer's state.
+type Observer interface {
+	// Range iterates over all events visible to the Observer. The caller is
+	// responsible for invoking Observe as part of the provided callback. Range
+	// can be used to customize setup and teardown behavior if the Observer
+	// wraps another Observer.
+	Range(func(watch.Event) bool)
```

Comment:
Given the bit of awkwardness [with All](https://github.com/pulumi/pulumi-kubernetes/pull/3134/files#diff-2042e56949b8976fc34e1d68bb8c4f9ed01d7f9b5796392b4893ba77e0f5f416R63-R65), would it make sense to factor out the `Range` function, e..g to source (the observed thing).